### PR TITLE
perf: improve regex performance with length limitation

### PR DESCRIPTION
**Pull request by Bingjiling**
Issue #, if available:

Description of changes:

* Add length limitation for regex to address CodeQL scan discovery 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/router/bundle/bundleParser.test.ts [hunk 1]
```diff
@@ -80,6 +80,71 @@ describe('parseResource', () => {
...
+                ],
+            };
+
+            try {
+                // OPERATE
+                await BundleParser.parseResource(bundleRequestJson, DynamoDbDataService, serverUrl);
+            } catch (e) {
+                // CHECK
+                expect((e as any).name).toEqual('Error');
+                expect((e as any).message).toEqual('Entry full URL length exceeds length limit of 500');
```
Comment (ssvegaraju) [inline src/router/bundle/bundleParser.test.ts]: Maybe we shouldn't hardcode the limit value here, either we could use the CONST value directly or just say it exceeds the length limit here. I think the latter is a better approach as it gives less information to any malicious attackers. Same for the above test on reference URL

File: src/router/bundle/bundleParser.ts [hunk 2]
```diff
@@ -334,6 +336,9 @@ export default class BundleParser {
 
         const references: Reference[] = referencePaths.map((referencePath) => {
             const entryReference = get(entry.resource, referencePath);
+            if (entryReference.length > MAX_REFERENCE_URL_LENGTH) {
+                throw new Error(`Reference URL exceeds length limit of ${MAX_REFERENCE_URL_LENGTH}`);
```
Comment (ssvegaraju) [inline src/router/bundle/bundleParser.ts]: same nitpick as above with including the length limit value

**Review by ssvegaraju — COMMENTED** [review]
LGTM! Just a minor comment:

**Review by ssvegaraju — DISMISSED** [review]

File: src/regExpressions.test.ts [hunk 3]
```diff
@@ -64,3 +70,41 @@ describe('captureFullUrlParts', () => {
...
+    test('Success match', () => {
+        const resourceString = `Patient/12345678`;
+        const actualMatch = resourceString.match(captureResourceTypeRegExp);
+
+        const expectedMatch = ['Patient/12345678', 'Patient'];
+
+        for (let i = 0; i < expectedMatch.length; i += 1) {
+            // @ts-ignore
+            expect(actualMatch[i]).toEqual(expectedMatch[i]);
+        }
```
Comment (nguyen102) [inline src/regExpressions.test.ts:84]: Couldn't you just do `expect(actualMatch).toEqual(expectedMatch)`

File: src/router/bundle/bundleParser.ts [hunk 4]
```diff
@@ -353,9 +358,14 @@ export default class BundleParser {
                 let rootUrl = fullUrlMatch[1];
                 // If the reference doesn't have a urlRoot, check if the entry's fullUrl has a urlRoot
                 if (rootUrl === undefined) {
-                    if (entry.fullUrl && entry.fullUrl.match(captureFullUrlParts)) {
+                    if (entry.fullUrl) {
```
Comment (nguyen102) [inline src/router/bundle/bundleParser.ts]: Thoughts on avoiding the deeply nested if statement by breaking the code into two if statements
```
if (rootUrl === undefined && entry.fullUrl && entry.fullUrl.length > MAX_BUNDLE_ENTRY_URL_LENGTH) {  // throw error} 
elseif (rootUrl === undefined && enty.fullUrl) {
 rootUrl = entry.fullUrl.match(captureFullUrlParts)
                            ? entry.fullUrl.match(captureFullUrlParts)[1]
                            : undefined;
}
```

File: src/regExpressions.test.ts (hunk 3 above)
Comment (Bingjiling) [reply src/regExpressions.test.ts:84]: The actualMatch is actually longer than the expectedMatch, with additional metadata fields like 'input', and those are the fields that we don't care.

Comment (nguyen102) [reply src/regExpressions.test.ts:84]: Ah, can we use slice then?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

`expect(actualMatch).toEqual(expectedMatch.slice(expectedMatch.length))`

File: src/router/bundle/bundleParser.ts (hunk 4 above)
Comment (Bingjiling) [reply src/router/bundle/bundleParser.ts]: Seems like CodeQL will still flag it if the two clause are separated completely. I took a middle path of organizing it to two layer of if statement.

File: src/regExpressions.test.ts (hunk 3 above)
Comment (Bingjiling) [reply src/regExpressions.test.ts:84]: Emmm that could work. I copied the for loop check from other test and didn't update it.

Comment (Bingjiling) [reply src/regExpressions.test.ts:84]: Emmm it didn't work actually. The resulted actualMatch result is of type RegExpMatchArray, which is extended from Array. But surprisingly slice does not work on it. I'll keep it as is.

File: src/regExpressions.test.ts [hunk 5]
```diff
@@ -64,3 +70,41 @@ describe('captureFullUrlParts', () => {
...
+    });
+    test('Extra long resource type fail to match', () => {
+        const resourceString = `${'Patient'.repeat(30)}/12345678`;
+        console.log(resourceString);
```
Comment (nguyen102) [inline src/regExpressions.test.ts]: Should this be deleted?

File: src/regExpressions.test.ts [hunk 6]
```diff
@@ -64,3 +70,41 @@ describe('captureFullUrlParts', () => {
...
+        const expectedMatch = ['Patient/12345678', 'Patient'];
+
+        for (let i = 0; i < expectedMatch.length; i += 1) {
+            // @ts-ignore
+            expect(actualMatch[i]).toEqual(expectedMatch[i]);
+        }
+    });
+    test('Extra long resource id fail to match', () => {
+        const resourceString = `Patient/${'12345678'.repeat(14)}`;
+        console.log(resourceString);
```
Comment (nguyen102) [inline src/regExpressions.test.ts]: should this be deleted?

**Review by nguyen102 — APPROVED** [review]

**Pull request merged**